RFC 53: updates to RFC process (#336)
aws/aws-cdk-rfcs: README.md

guidance.

### 1. Create a tracking issue

Each RFC has a GitHub issue that tracks it from start to finish. The issue is
the hub for conversations, community signal (+1s) and the issue number is used
as the unique identifier of this RFC.

> Before creating a tracking issue, please search for similar or related ideas in
the RFC table above or in the issue list of this repo. If there is a relevant
RFC, collaborate on that existing RFC, based on its current stage.

Our [tracking issue template] includes a checklist of all the steps an RFC goes
through and it's the driver's responsibility to update the checklist and assign
the correct label to on the RFC throughout the process.

[tracking issue template]: https://github.com/aws/aws-cdk-rfcs/blob/master/.github/ISSUE_TEMPLATE/tracking-issue.md

When the issue is created, it is required to fill in the following information:

1. **Title**: the name of the feature or change - think changelog entry.
2. **Description**: a _short_ description of feature, as if it was already implemented.
3. **Proposed by**: fill in the GitHub alias of the person who proposed the idea
   under "Proposed by".

> ### 2. Identify roles and responsibilities

Reviewer: As mentioned above, we should describe how to find people for these roles.

Moreover, the first part of this README says that 'anyone' can create an RFC. If so, we need a lot more guidance for contributors not on the team.

.... Or, we reduce the scope and say that RFCs are created by the CDK team. (we would still need some guidance in this paragraph)

Author: Modified to reframe as "API Bar Raiser" - let me know what do you think

Each RFC has a __driver__, __approver(s)__ and __stakeholders__.

The __driver__ is the person responsible to drive this RFC to completion. This
does __not__ mean that the driver must be the person actually doing all the
work, but it is the person responsible to get the RFC through the finish line.

__Approvers__ are the people that have the authority to approve the RFC and any
subsequent changes to it. Approvers have the final word if we cannot reach
consensus.

__Stakeholders__ are people (internal/external) who may be impacted by this RFC
or work on similar domains or projects. The goal is to make sure that
stakeholders have an opportunity to influence the direction of the RFC in
various stages of the process.

The driver is responsible to update the **Roles** table in the tracking issue.

> ### 3. Organize a kick-off meeting

Reviewer: Similar comment as above.

> Before diving into writing the RFC, is it highly recommended to organize a

Reviewer: ```suggestion
Before diving into writing the RFC, it is highly recommended to organize a
```

kick-off meeting that includes the driver, approver(s) and stakeholders. The
goal of the meeting is to discuss the feature, its scope and general direction
for implementation.

> Our experience shows that such a meeting can save a lot of time and energy.

Reviewer: My suggestion would be to do something different. It would be to identify 'some' high level approaches, put that down in the RFC or tracking issue, and then have that kickoff meeting. 

This way, there is some starting point and people don't arrive at a meeting with a blank slate.

Author: I'll add a recommendation to write something up in the tracking issue as a prep for that meeting

~week, if no major concerns are raised, the RFC will be approved and merged.

Add a comment on the RFC pull request and tracking issue that the RFC entered
this stage so that all relevant stakeholders will be notified.

### 8. Obtain final approval

Once the final comments period is over, and the author has made final edits, the
RFC can be approved by the approvers and merged into the repository.

NOTE: only RFCs that were approved by approvers should be merged.

### 9. Implementation

For large changes, we highly recommend creating an implementation plan which
lists all the tasks required. In many cases, large implementation  should be
broken down and released via multiple iterations. Devising a concrete plan to
break down the break can be very helpful.

The implementation plan should be submitted through a PR that adds an addendum
to the RFC document and seeks the approval of any relevant stakeholders.

Throughout this process, update the tracking issue:

- Add the alias of the "implementation lead"
- Execution plan submitted (label: `status/planning`)
- Plan approved and merged (label: `status/implementing`)
- Implementation complete (label: `status/done`)

> ## State Diagram

Reviewer: This section needs updating as well.

Author: Not sure - it's generally the same state flow

## What is an RFC?

An RFC is a document that proposes a change to one of the projects led by the
CDK team at AWS. *Request for Comments* means a request for discussion and
oversight about the future of the project from maintainers, contributors and
users.

> **When to submit an RFC?** Anyone can submit an RFC for *any* change, big or

Author: add: The goal of the RFC is to reduce risk and ambiguity. One question that you should be able to answer after an RFC is ready is how long will it take to implement this in production quality. Do you both feel you flushed out enough of the design to be able to provide a confident answer?

Author: Added

[tracking issue template]: https://github.com/aws/aws-cdk-rfcs/blob/master/.github/ISSUE_TEMPLATE/tracking-issue.md

When the issue is created, it is required to fill in the following information:

1. **Title**: the name of the feature or change - think changelog entry.
2. **Description**: a _short_ description of feature, as if it was already implemented.
3. **Proposed by**: fill in the GitHub alias of the person who proposed the idea
   under "Proposed by".

### 2. API Bar Raiser

The public API of a feature represents the surface through which users interact
with it, and we want to make sure these APIs are consistent, ergonomic and
designed based on the intent and the mental model of our users. Additionally,
once we announce that a feature is "stable" (1.0, GA, etc) any breaking change
to its public API will require releasing a new major version, so we think of API
decisions as "one way doors".

For each RFC, the core team will assign an **API Bar Raiser** who reviews and
approves any public API of the feature. API Bar Raisers have veto rights on
API-related design decisions, such as naming, structure, options, CLI commands
and others.

> After an RFC is proposed, the core team will assign it an API Bar Raiser using a

Reviewer: The definition core team is changing quite a bit with the two teams, etc. Should we say 'cdk leads' instead?

Author: Fixed

> tiering model generally based on the size of the user base that will likely get

Reviewer: Where do you plan to document the tiering model and how they're determined?

Author: Our internal team flow documents

> exposed to the feature. As a general rule, the more "significant" the change is,

Reviewer: ```suggestion
exposed to the feature. As a general rule, for rfcs with "significant" changes,
```

Author: rephrased a bit

we will assign a bar raiser with a wider and longer-term context of the project.

API Bar Raisers are required to **sign off** on the **Working Backwards**
section of your RFC, so we encourage to engage with them early in the process to
make sure you are aligned on how the API should be designed.

The API Bar Raiser is called out in the RFC tracking issue and in the RFC
document itself. Feel free to reach us at the

> [#aws-cdk-rfcs](https://cdk-dev.slack.com/archives/C025ZFGMUCD) channel in the

Reviewer: We have not established a regular practice of monitoring this. It shouldn't rely on something one person does.

We should include this as part of our regular 'triage' process, included into our usual monitoring to ensure incoming issues get triaged.

Author: I'll add to team flow

Author: https://github.com/cdklabs/team-internal/pull/18

[cdk.dev](https://cdk.dev) slack workspace if you need a bar raiser assigned to
your RFC.

> The engineering solution proposed in an RFC does not need require approval
> beyond the normal pull request approval model (e.g. a core team member needs
> to approve the RFC PR and any subsequent changes to it).

### 3. Organize a kick-off meeting

Before diving into writing the RFC, it is highly recommended to organize a
kick-off meeting that includes the API Bar Raiser and any stakeholders that
might be interested in this RFC or can contribute ideas and direction. The goal
of the meeting is to discuss the feature, its scope and general direction for
implementation.

Our experience shows that such a meeting can save a lot of time and energy.

Use the RFC tracking issue to record some initial API and design ideas and

> collect early feedback and use cases as a preparation for the kick off meeting.

Reviewer: I generally prefer doing this in the RFC rather than tracking issue. Basically a rough half baked (or not baked) rfc. 

It allows me to iterate on my idea via git, before I'm ready to share.

Can we say either or move to that model? I generally find doing this on issues to be a little hard to manage.

Author: I think some initial iteration on the issue is desirable (and actually came up in our retrospective as a light-weight process to get started).

Author: Changed to "you can..."

A few tips:

- If you decide to resolve a comment without addressing it, take the time to
  explain.
- Try to understand where people are coming from. If a comment seems off, ask
  folks to elaborate and describe their use case or provide concrete examples.
- Work with your API bar raiser: if there are disagreements, @mention them in a
  comment and ask them to provide their opinion.
- Be patient: it sometimes takes time for an RFC to converge. Our experience
  shows that some ideas need to "bake" and solutions oftentimes emerge via a
  healthy debate. We've had RFCs that took months to resolve.
- Not everything must be resolved in the first revision. It is okay to leave
  some things to resolve later. Make sure to capture them clearly and have an
  agreement about that. We oftentimes update an RFC doc a few times during the
  implementation.

### 7. API Bar Raiser Sign-off

Before you can merge your RFC, you will need the API Bar Raiser to sign off on
the **Working Backwards** section of your document.

To record this sign-off, the bar raiser will add a comment to your RFC PR
indicating that the working backwards section looks good to them (LGTM) and will

> update the RFC doc to check the "sign off" box in the working backwards section

Reviewer: why not simply add a label?

Author: Good call

Author: Fixed.

to its public API will require releasing a new major version, so we like think
of API decisions as "one way doors".

API Bar Raisers will be assigned using a tiering model which is generally based
on the size of the user base that will likely get exposed to the feature. As a
general rule, the more "significant" the feature is, we will assign a bar raiser
with a wider and longer-term context of the project.

To merge an RFC, a [sign-off](#6-api-sign-off) from the bar raiser is required
on the public API of the feature, so we encourage to engage with them early in
the process to make sure you are aligned on how the API should be designed.

> NOTE: The technical solution proposed in an RFC *does not* require approval
> beyond the normal pull request approval model (e.g. a core team member needs
> to approve the RFC PR and any subsequent changes to it).

### 3. Kick-off

Before diving into writing the RFC, it is highly recommended to organize a
kick-off meeting that includes the API Bar Raiser and any stakeholders that
might be interested in this RFC or can contribute ideas and direction. The goal
of the meeting is to discuss the feature, its scope and general direction for

> implementation.

Reviewer 2: Perhaps suggest how to go about this? From an external collaborator's perspective, it is not clear how to get a hold of people's calendars, what conferencing solutions are available - which are free to use? which are acceptable?

Author: Good call. I'll add a suggestion to reach out to the team via the #aws-cdk-rfcs slack channel to get help organizing this

Author: Added

designed based on the intent and the mental model of our users. Additionally,
once we announce that a feature is "stable" (1.0, GA, etc) any breaking change
to its public API will require releasing a new major version, so we think of API
decisions as "one way doors".

For each RFC, the core team will assign an **API Bar Raiser** who reviews and
approves any public API of the feature. API Bar Raisers have veto rights on
API-related design decisions, such as naming, structure, options, CLI commands
and others.

After an RFC is proposed, the core team will assign it an API Bar Raiser using a
tiering model generally based on the size of the user base that will likely get
exposed to the feature. As a general rule, the more "significant" the change is,
we will assign a bar raiser with a wider and longer-term context of the project.

API Bar Raisers are required to **sign off** on the **Working Backwards**
section of your RFC, so we encourage to engage with them early in the process to
make sure you are aligned on how the API should be designed.

The API Bar Raiser is called out in the RFC tracking issue and in the RFC

> document itself. Feel free to reach us at the

Reviewer 3: I'm confused about whether the bar raiser is assigned by the team or the author should reach out to get one assigned

Author: Either option is good. This is targeting both internal and external folks, so I wanted to give both options.
